ludwig/decoders/llm_decoders.py

```diff
@@ -0,0 +1,133 @@
+import logging
+import re
+from typing import Any, Dict, Union
+
+import torch
+
+from ludwig.api_annotations import DeveloperAPI
+from ludwig.constants import CATEGORY
+from ludwig.decoders.base import Decoder
+from ludwig.decoders.registry import register_decoder
+
+# from ludwig.schema.features.preprocessing.category import CategoryOutputPreprocessingConfig
+from ludwig.schema.decoders.llm_decoders import ParserDecoderConfig
+from ludwig.utils.strings_utils import get_tokenizer
+
+logger = logging.getLogger(__name__)
+
+
+# TODO(Arnav): Refactor to split into strategies like splitters
+class Matcher:
+    def __init__(self, match: Dict[str, Dict[str, Any]]):
+        self.match = match
+
+    def contains(self, decoded_input: str, value: str) -> bool:
+        return value in decoded_input
+
+    def regex(self, decoded_input: str, regex_pattern: str) -> bool:
+        """Perform a regex match on a given text using a specified regex pattern.
+
+        Parameters:
+        text (str): The text to perform the match on.
+        regex_pattern (str): The regex pattern to use for the match.
+
+        Returns:
+        A list of match objects.
+        """
+        # Compile the regex pattern
+        matches = []
+        try:
+            regex = re.compile(regex_pattern)
+            # Perform the match
+            matches = regex.findall(decoded_input)
+        except Exception:
+            logger.warning(f"Regex pattern {regex_pattern} could not be compiled.")
+        # If there is a match, matches is a non-empty list, so we can use this
+        # to infer if there was a match or not and return a bool
+        return len(matches) > 0
+
+    def __call__(self, decoded_input: str) -> Union[str, None]:
+        # Greedy match on first label that matches the input
+        for label, label_def in self.match.items():
+            label_def_type = label_def["type"]
+            label_def_value = label_def["value"]
+
+            if label_def_type == "contains":
+                is_match = self.contains(decoded_input, label_def_value)
+            elif label_def_type == "regex":
+                is_match = self.regex(decoded_input, label_def_value)
+            else:
+                raise ValueError(
+                    f"{label_def_type} is not a valid match `type`. Ludwig "
+                    "currently supports `contains` and `regex` match types."
+                )
+
+            if is_match:
+                return label
+        return None
+
+
+@DeveloperAPI
+@register_decoder("parser", [CATEGORY])
+class ParserDecoder(Decoder):
```

Comment:
it's an open question to me how to call this thing. i understand why parser, but it's not really the decoder that is a parser, it's the extractor function after the decoding. We can keep it for now but let's discuss about it

Reply:
That's true.

I've also changed this now to be `category_parser` and `text_parser`, so we should probably think through it further. One potential name change could be "Extractor" instead of "Parser", since that is technically what the forward pass of both decoders are doing. Happy to change it to this if you think it is better.